helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/AssignableNode.java

```diff
@@ -19,10 +19,293 @@
+    return _maxCapacity;
+  }
+
+  public int getMaxPartition() {
+    return _maxPartition;
+  }
+
+  /**
+   * Computes the fault zone id based on the domain and fault zone type when topology is enabled. For example, when
+   * the domain is "zone=2, instance=testInstance" and the fault zone type is "zone", this function returns "2".
+   * If cannot find the fault zone id, this function leaves the fault zone id as the instance name.
+   *
+   * TODO merge this logic with Topology.java tree building logic.
+   * For now, the WAGED rebalancer has a more strict topology def requirement.
+   * Any missing field will cause an invalid topology config exception.
+   */
+  private String computeFaultZone(ClusterConfig clusterConfig, InstanceConfig instanceConfig) {
+    if (clusterConfig.isTopologyAwareEnabled()) {
+      String topologyStr = clusterConfig.getTopology();
+      String faultZoneType = clusterConfig.getFaultZoneType();
+      if (topologyStr == null || faultZoneType == null) {
+        throw new HelixException("Fault zone or cluster topology information is not configured.");
+      }
+
+      String[] topologyDef = topologyStr.trim().split("/");
+      if (topologyDef.length == 0 || Arrays.stream(topologyDef)
+          .noneMatch(type -> type.equals(faultZoneType))) {
+        throw new HelixException(
+            "The configured topology definition is empty or does not contain the fault zone type.");
+      }
+
+      Map<String, String> domainAsMap = instanceConfig.getDomainAsMap();
+      if (domainAsMap == null) {
+        throw new HelixException(String
+            .format("The domain configuration of instance %s is not configured", _instanceName));
+      } else {
+        StringBuilder faultZoneStringBuilder = new StringBuilder();
+        for (String key : topologyDef) {
+          if (!key.isEmpty()) {
+            if (domainAsMap.containsKey(key)) {
+              faultZoneStringBuilder.append(domainAsMap.get(key));
+              faultZoneStringBuilder.append('/');
+            } else {
+              throw new HelixException(String.format(
+                  "The domain configuration of instance %s is not complete. Type %s is not found.",
+                  _instanceName, key));
+            }
+            if (key.equals(faultZoneType)) {
+              break;
+            }
+          }
+        }
+        return faultZoneStringBuilder.toString();
+      }
+    } else {
+      // For backward compatibility
+      String zoneId = instanceConfig.getZoneId();
+      return zoneId == null ? instanceConfig.getInstanceName() : zoneId;
+    }
+  }
+
+  /**
+   * This function should only be used to assign a set of partitions that doesn't exist before.
+   * Using this function avoids the overhead of updating capacity repeatedly.
+   */
+  private void assignNewBatch(Collection<AssignableReplica> replicas) {
+    Map<String, Integer> totalPartitionCapacity = new HashMap<>();
+    for (AssignableReplica replica : replicas) {
```

Comment:
Is there a deterministic order to this set of replicas? Let us make this crystal clear because stateful systems might care about the order and may want the flexibility to assign priority among new replicas?

Reply:
This is intentionally done. The algorithm should not rely on the order in cluster model to calculate. Since different algorithms can have a different requirement.

Reply:
I agree with @jiajunwang. We shouldn't make any assumptions about the order of the incoming parameter nor the returned collections. Unless the code itself explicitly set so, like using `ImmutableList`. Writing comments and hope others will respect the comment just makes our lives harder.